lib/util/cmd.go

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+package util
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/adobe/aquarium-fish/lib/log"
+)
+
+// Runs & logs the executable command
+func RunAndLog(sec string, timeout time.Duration, stdin io.Reader, path string, arg ...string) (string, string, error) {
+	var stdout, stderr bytes.Buffer
+
+	// Running command with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, path, arg...)
+
+	log.Debugf("%s: Executing: %s %s", sec, cmd.Path, strings.Join(cmd.Args[1:], " "))
+	if stdin != nil {
+		cmd.Stdin = stdin
+	}
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+
+	stdoutString := strings.TrimSpace(stdout.String())
+	stderrString := strings.TrimSpace(stderr.String())
+
+	// Check the context error to see if the timeout was executed
+	if ctx.Err() == context.DeadlineExceeded {
+		err = fmt.Errorf("%s: Command timed out", sec)
+	} else if _, ok := err.(*exec.ExitError); ok {
+		message := stderrString
+		if message == "" {
+			message = stdoutString
+		}
+
+		err = fmt.Errorf("%s: Command exited with error: %v: %s", sec, err, message)
+	}
+
+	if len(stdoutString) > 0 {
+		log.Debugf("%s: stdout: %s", sec, stdoutString)
+	}
+	if len(stderrString) > 0 {
+		log.Debugf("%s: stderr: %s", sec, stderrString)
+	}
+
+	// Replace these for Windows, we only want to deal with Unix style line endings.
+	returnStdout := strings.ReplaceAll(stdout.String(), "\r\n", "\n")
+	returnStderr := strings.ReplaceAll(stderr.String(), "\r\n", "\n")
+
+	return returnStdout, returnStderr, err
+}
+
+// Will retry on error and store the retry output and errors to return
+func RunAndLogRetry(sec string, retry int, timeout time.Duration, stdin io.Reader, path string, arg ...string) (stdout string, stderr string, err error) {
+	counter := 0
+	for {
+		counter++
+		rout, rerr, err := RunAndLog(sec, timeout, stdin, path, arg...)
+		if err != nil {
+			stdout += fmt.Sprintf("\n--- %s: Command execution attempt %d ---\n", sec, counter)
+			stdout += rout
+			stderr += fmt.Sprintf("\n--- %s: Command execution attempt %d ---\n", sec, counter)
+			stderr += rerr
+			if counter <= retry {
+				// Give command time to rest
+				time.Sleep(time.Duration(counter) * time.Second)
+				continue
```

Comment:
I'm a big fan of this code getting consolidated into one module.  I noticed that there were two versions, one that does `time.Duration(counter) * time.Second`, and a hard-coded version

```go
// Give command 5 seconds to rest
time.Sleep(5 * time.Second)
continue
```

Most of the use sites are asking for `3` retries, meaning you wait `1` second, then `2` seconds, then `3` seconds.  It seems like if I asked for `30` retries, this behavior of waiting an increasing amount of time is not desirable in favor of just blanket "wait 5 seconds every time."

Was this the version you intended to keep?

Reply:
I do like the increasing timer - because it allows to do retry faster in the beginning then 5 sec, it's important when your app init needs to retry for example and 150 vs 465 sec is quite a big difference, but we don't use 30 retries in the logic and anyone who want - will need to make sure he understands the logic and how much it will wait. Maybe in the future we will swich to timeout and retries within this timeout - but for now for 5-10 retries it works relatively well.

Reply:
Works for me, I can see value to either version.  Just wanted to make sure you were using the 1/2 implementations you wanted in the final shared version :slightly_smiling_face: